src/sfall_arrays.cc

```diff
@@ -0,0 +1,697 @@
+            MapSort(newLen);
+        }
+    }
+
+    ProgramValue ScanArray(const ProgramValue& value, Program* program)
+    {
+        auto valueEl = ArrayElement { value, program };
+        auto it = std::find_if(pairs.begin(), pairs.end(), [&valueEl](const KeyValuePair& pair) {
+            return pair.value == valueEl;
+        });
+
+        if (it == pairs.end()) {
+            return ProgramValue(-1);
+        }
+
+        return it->key.toValue(program);
+    }
+
+private:
+    struct KeyValuePair {
+        ArrayElement key;
+        ArrayElement value;
+    };
+
+    void MapSort(int type)
+    {
+        bool sortByValue = false;
+        if (type < ARRAY_ACTION_SHUFFLE) {
+            type += 4;
+            sortByValue = true;
+        }
+
+        if (sortByValue) {
+            ListSort(pairs, type, [](const KeyValuePair& a, const KeyValuePair& b) -> bool {
+                return a.value < b.value;
+            });
+        } else {
+            ListSort(pairs, type, [](const KeyValuePair& a, const KeyValuePair& b) -> bool {
+                return a.key < b.key;
+            });
+        }
+    }
+
+    std::vector<KeyValuePair> pairs;
+};
+
+struct SfallArraysState {
+    std::unordered_map<ArrayId, std::unique_ptr<SFallArray>> arrays;
+    std::unordered_set<ArrayId> temporaryArrayIds;
+    int nextArrayId = kInitialArrayId;
+};
+
+static SfallArraysState* _state = nullptr;
+
+bool sfallArraysInit()
+{
+    _state = new (std::nothrow) SfallArraysState();
+    if (_state == nullptr) {
+        return false;
+    }
+
+    return true;
+}
+
+void sfallArraysReset()
+{
+    if (_state != nullptr) {
+        _state->arrays.clear();
+        _state->temporaryArrayIds.clear();
+        _state->nextArrayId = kInitialArrayId;
+    }
+}
+
+void sfallArraysExit()
+{
+    if (_state != nullptr) {
+        delete _state;
+    }
+}
+
+ArrayId CreateArray(int len, unsigned int flags)
+{
+    flags = (flags & ~1); // reset 1 bit
+
+    if (len < 0) {
+        flags |= SFALL_ARRAYFLAG_ASSOC;
+    } else if (len > ARRAY_MAX_SIZE) {
+        len = ARRAY_MAX_SIZE; // safecheck
+    }
+
+    ArrayId arrayId = _state->nextArrayId++;
+
+    if (flags & SFALL_ARRAYFLAG_ASSOC) {
+        _state->arrays.emplace(std::make_pair(arrayId, std::make_unique<SFallArrayAssoc>(flags)));
+    } else {
+        _state->arrays.emplace(std::make_pair(arrayId, std::make_unique<SFallArrayList>(len, flags)));
+    }
+
+    return arrayId;
+}
+
+ArrayId CreateTempArray(int len, unsigned int flags)
+{
+    ArrayId arrayId = CreateArray(len, flags);
+    _state->temporaryArrayIds.insert(arrayId);
+    return arrayId;
+}
+
+SFallArray* get_array_by_id(ArrayId arrayId)
+{
+    auto it = _state->arrays.find(arrayId);
+    if (it == _state->arrays.end()) {
+        return nullptr;
+    }
+
+    return it->second.get();
+}
+
+ProgramValue GetArrayKey(ArrayId arrayId, int index, Program* program)
+{
+    auto arr = get_array_by_id(arrayId);
+    if (arr == nullptr) {
```

Comment:
As far as I know build-in pointer-to-bool conversion works exactly this way. Just a matter of style